src/common/meta/src/key/tombstone.rs

```diff
@@ -0,0 +1,448 @@
+
+/// [TombstoneManager] provides the ability to:
+/// - logically delete values
+/// - restore the deleted values
+pub(crate) struct TombstoneManager {
+    kv_backend: KvBackendRef,
+}
+
+const TOMBSTONE_PREFIX: &str = "__tombstone/";
+
+pub(crate) struct TombstoneKey<T>(T);
+
+fn to_tombstone(key: &[u8]) -> Vec<u8> {
+    [TOMBSTONE_PREFIX.as_bytes(), key].concat()
+}
+
+impl TombstoneKey<&Vec<u8>> {
+    /// Returns the origin key and tombstone key.
+    fn to_keys(&self) -> (Vec<u8>, Vec<u8>) {
+        let key = self.0;
+        let tombstone_key = to_tombstone(key);
+        (key.clone(), tombstone_key)
+    }
+
+    /// Returns the origin key and tombstone key.
+    fn into_keys(self) -> (Vec<u8>, Vec<u8>) {
+        self.to_keys()
+    }
+
+    /// Returns the tombstone key.
+    fn to_tombstone_key(&self) -> Vec<u8> {
+        let key = self.0;
+        to_tombstone(key)
+    }
+}
+
+impl TableMetaKeyGetTxnOp for TombstoneKey<&Vec<u8>> {
+    fn build_get_op(
+        &self,
+    ) -> (
+        TxnOp,
+        impl FnMut(&'_ mut TxnOpGetResponseSet) -> Option<Vec<u8>>,
+    ) {
+        let key = to_tombstone(self.0);
+        (TxnOp::Get(key.clone()), TxnOpGetResponseSet::filter(key))
+    }
+}
+
+/// Atomic Key:
+/// The value corresponding to the key remains consistent between two transactions.
+pub(crate) enum Key {
+    Atomic(Vec<u8>),
+    Other(Vec<u8>),
+}
+
+impl Key {
+    /// Returns a new [Key::Atomic].
+    pub(crate) fn atomic<T: Into<Vec<u8>>>(key: T) -> Self {
+        Self::Atomic(key.into())
+    }
+
+    /// Returns a new [Key::Other].
+    pub(crate) fn other<T: Into<Vec<u8>>>(key: T) -> Self {
+        Self::Other(key.into())
+    }
+
+    fn get_inner(&self) -> &Vec<u8> {
+        match self {
+            Key::Atomic(key) => key,
+            Key::Other(key) => key,
+        }
+    }
+
+    fn is_atomic(&self) -> bool {
+        matches!(self, Key::Atomic(_))
+    }
+}
+
+impl TableMetaKeyGetTxnOp for Key {
+    fn build_get_op(
+        &self,
+    ) -> (
+        TxnOp,
+        impl FnMut(&'_ mut TxnOpGetResponseSet) -> Option<Vec<u8>>,
+    ) {
+        let key = self.get_inner().clone();
+        (TxnOp::Get(key.clone()), TxnOpGetResponseSet::filter(key))
+    }
+}
+
+impl TombstoneManager {
+    /// Returns [TombstoneManager].
+    pub fn new(kv_backend: KvBackendRef) -> Self {
+        Self { kv_backend }
+    }
+    /// Creates tombstones for keys.
+    ///
+    /// Preforms to:
+    /// - retrieve all values corresponding `keys`.
+    /// - stores tombstone values.
+    pub(crate) async fn create(&self, keys: Vec<Key>) -> Result<bool> {
+        // Builds transaction to retrieve all values
+        let (operations, mut filters): (Vec<_>, Vec<_>) =
+            keys.iter().map(|key| key.build_get_op()).unzip();
+
+        let txn = Txn::new().and_then(operations);
+        let mut resp = self.kv_backend.txn(txn).await?;
+        ensure!(
+            resp.succeeded,
+            error::UnexpectedSnafu {
```

Comment:
Better carry more info in the error's output, otherwise it's hard to see why the error is occurred.

Reply:
Good point🚀

Reply:
Can't see where it's resolved?